public/pages/workflow_detail/workspace/workspace.tsx

```diff
@@ -3,24 +3,161 @@
-import { AppState } from '../../../store';
-import { WorkspaceComponent } from '../workspace_component';
+import { EuiFlexItem, EuiFlexGroup } from '@elastic/eui';
+import { AppState, rfContext } from '../../../store';
+import { convertToReactFlowData } from '../../../../common';
 
-export function Workspace() {
-  const { components } = useSelector((state: AppState) => state.workspace);
+// styling
+import 'reactflow/dist/style.css';
+import './reactflow-styles.scss';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface WorkspaceProps {}
+
+export function Workspace(props: WorkspaceProps) {
+  const reactFlowWrapper = useRef(null);
+  const { reactFlowInstance, setReactFlowInstance } = useContext(rfContext);
+
+  // Fetching workspace state to populate the initial nodes/edges.
+  // Where/how the low-level ReactFlow JSON will be persisted is TBD.
+  // TODO: update when that design is finalized
+  const storedComponents = useSelector(
+    (state: AppState) => state.workspace.components
+  );
+  const { rfNodes, rfEdges } = convertToReactFlowData(storedComponents);
+  const [nodes, setNodes, onNodesChange] = useNodesState(rfNodes);
+  const [edges, setEdges, onEdgesChange] = useEdgesState(rfEdges);
+
+  const onConnect = useCallback(
+    (params) => {
+      setEdges((eds) => addEdge(params, eds));
+    },
+    // TODO: add customized logic to prevent connections based on the node's
+    // allowed inputs. If allowed, update that node state as well with the added
+    // connection details.
+    [setEdges]
+  );
+
+  const onDragOver = useCallback((event) => {
+    event.preventDefault();
+    event.dataTransfer.dropEffect = 'move';
+  }, []);
+
+  const onDrop = useCallback(
+    (event) => {
+      event.preventDefault();
+      // Get the node info from the event metadata
+      const nodeData = event.dataTransfer.getData('application/reactflow');
+
+      // check if the dropped element is valid
+      if (typeof nodeData === 'undefined' || !nodeData) {
+        return;
+      }
+
+      // Fetch bounds based on the ref'd div component, adjust as needed.
+      // TODO: remove hardcoded bounds and fetch from a constant somewhere
+      // @ts-ignore
+      const reactFlowBounds = reactFlowWrapper.current.getBoundingClientRect();
+      // @ts-ignore
+      const position = reactFlowInstance.project({
+        x: event.clientX - reactFlowBounds.left - 80,
+        y: event.clientY - reactFlowBounds.top - 90,
+      });
+
+      // TODO: remove hardcoded values when more component info is passed in the event.
+      // Only keep the calculated 'positioning' field.
+      const newNode = {
+        // TODO: generate ID based on the node data maybe
+        id: Date.now().toFixed(),
+        type: nodeData.type,
+        position,
+        data: { label: nodeData.label },
+        style: {
+          background: 'white',
+        },
+      };
+
+      // TODO: add logic to add node into the redux datastore
```

Comment:
I thought we aren't storing the nodes in the redux while updating the workflow and just the hook here? Is that in-correct. We will update a redux datastore as we drop in nodes?

Reply:
Thanks for the catch, yes you're correct. We will utilize reactFlowContext and useNodesState() / useEdgesState() to maintain all of this for us. I'll remove the comment